### PR TITLE
Improvements to syntax handling for data types

### PR DESCRIPTION
**Pull request by niklasharrysson**
- Added new base class for type syntax handling with default implementation for scalar and aggregate types
- Added derived classes overriding the syntax where needed in GLSL and OSL
- Fixed a long-standing bug with OSL syntax for color4 type, now possible to fix with the new more flexible system
- Also removed support for swizzling by channel attribute, since this was removed in the specification v1.36

### CODE REVIEW

File: source/MaterialXShaderGen/SgNode.h [hunk 1]
```diff
@@ -33,7 +33,6 @@ class SgInput
     SgNode* node;
     ValuePtr value;
     SgOutput* connection;
-    string channels;
```
Comment (niklasharrysson) [inline source/MaterialXShaderGen/SgNode.h]: Remove the support for direct connection swizzling by a channel attribute

File: source/MaterialXShaderGen/Syntax.h [hunk 2]
```diff
@@ -12,95 +12,62 @@
...
-/// Base class for syntax objects used by code generators
-/// to emit code with correcy syntax for each language
+/// Base class for syntax objects used by shader generators
+/// to emit code with correcy syntax for each language.
 class Syntax
 {
 public:
-    /// Struct holding information on syntax to be used for data types
-    /// Required to be set for each data type
-    struct TypeSyntax
```
Comment (niklasharrysson) [inline source/MaterialXShaderGen/Syntax.h]: Pulling this out and making it a proper base class, with derived implementations for various type syntaxes.

File: source/MaterialXShaderGen/Syntax.h [hunk 3]
```diff
@@ -12,95 +12,62 @@
...
-            , defaultValue(dv)
-            , paramDefaultValue(pdv)
-            , typeDef(dn)
-            , outputName(on)
-        {}
-    };
-
-    /// Struct holding information on syntax needed to construct values and swizzle patterns for data types
-    /// Optional and only used if a custom constructor and/or swizzling is needed for a data type
-    struct ValueConstructSyntax
```
Comment (niklasharrysson) [inline source/MaterialXShaderGen/Syntax.h]: Removed this and merged its value construction functionality into the TypeSyntax class

File: source/MaterialXShaderGen/Shader.cpp [hunk 4]
```diff
@@ -72,7 +72,7 @@ void Shader::initialize(ElementPtr element, ShaderGenerator& shadergen, const Sg
                     // otherwise we can't assign user values to it anyway. 
                     // This is the case for BSDF, EDF, VDF, and shader types 
                     // which we never want to publish as editable uniforms.
-                    const string& typeDefault = shadergen.getSyntax()->getTypeDefault(input->type, true);
+                    const string& typeDefault = shadergen.getSyntax()->getDefaultValue(input->type, true);
```
Comment (bernardkwok) [inline source/MaterialXShaderGen/Shader.cpp]: Should this still be getTypeDefault vs getDefaultValue ?

File: source/MaterialXShaderGen/ShaderGenerators/Osl/OslSyntax.cpp [hunk 5]
```diff
@@ -1,12 +1,124 @@
...
+            }
+        }
+    };
+}
+
+const string OslSyntax::OUTPUT_QUALIFIER = "output";
+const vector<string> OslSyntax::VECTOR_MEMBERS  = { "[0]", "[1]", "[2]" };
+const vector<string> OslSyntax::VECTOR2_MEMBERS = { ".x", ".y" };
+const vector<string> OslSyntax::VECTOR4_MEMBERS = { ".x", ".y", ".z", ".w" };
+const vector<string> OslSyntax::COLOR2_MEMBERS  = { ".r", ".a" };
```
Comment (bernardkwok) [inline source/MaterialXShaderGen/ShaderGenerators/Osl/OslSyntax.cpp]: Guess this was what you were mentioning that we are stuck with different syntax for color vs vec.

File: source/MaterialXShaderGen/Shader.cpp (hunk 4 above)
Comment (niklasharrysson) [reply source/MaterialXShaderGen/Shader.cpp]: getTypeDefault was renamed to getDefaultValue for a more consistent naming and to better describe it's a value being returned.

File: source/MaterialXShaderGen/ShaderGenerators/Osl/OslSyntax.cpp (hunk 5 above)
Comment (niklasharrysson) [reply source/MaterialXShaderGen/ShaderGenerators/Osl/OslSyntax.cpp]: Yes, the whole syntax handling of types and swizzling of values would have been orders of magnitude simpler if OSL could just have 2- and 4- component vectors and colors as built in types, like any other shading language :(

**Pull request merged**